### PR TITLE
add awsCertlint worker

### PR DESCRIPTION
**Pull request by adamdecaf**
Fixes: https://github.com/mozilla/tls-observatory/issues/257 

```
# From tlsobs -observatory http://localhost:8083 maxx.nercomp.org 
--- Analyzers ---
* awslabs/certlint Results
 - No messages
```

```
# From tlsobs -observatory http://localhost:8083 federate.deere.com
--- Analyzers ---
* awslabs/certlint Results
 - Notice: Some python versions will not see SAN extension if it is the first extension	awslabs-certlint780452766
```

### CODE REVIEW

File: worker/awsCertlint/awsCertlint.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+	"github.com/mozilla/tls-observatory/certificate"
+	"github.com/mozilla/tls-observatory/worker"
+)
+
+var (
+	workerName = "awsCertlint"
+	workerDesc = "Runs awslabs/certlint over a given Certificate, categoriezes output for display on the certificate"
+
+	certlintDirectory = "/go/certlint" // path from tools/Dockerfile-scanner
```
Comment (adamdecaf) [inline worker/awsCertlint/awsCertlint.go]: This path isn't _ideal_, but is it worthwhile to change? Probably, since I assume we're wanting to add more binaries to check certificates with.

File: tools/Dockerfile-scanner [hunk 2]
```diff
@@ -4,28 +4,35 @@ ENV PROJECT=github.com/mozilla/tls-observatory
...
-    useradd -r -g tlsobs tlsobs 
+    useradd -r -g tlsobs tlsobs
 
 ADD . /go/src/$PROJECT
 ADD cipherscan/ /opt/cipherscan/
 ADD conf/ /etc/tls-observatory/
 RUN chown tlsobs -R /go /opt -R
 
 # switch to tlsobs user
+RUN mkdir -p /home/tlsobs && \
```
Comment (jvehent) [inline tools/Dockerfile-scanner]: maybe use `-m` in the useradd command instead of `mkdir /home/tlsobs` ?

File: worker/awsCertlint/awsCertlint.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	"os/exec"
+	"strings"
+
+	"github.com/mozilla/tls-observatory/certificate"
+	"github.com/mozilla/tls-observatory/worker"
+)
+
+var (
+	workerName = "awsCertlint"
+	workerDesc = "Runs awslabs/certlint over a given Certificate, categoriezes output for display on the certificate"
```
Comment (jvehent) [inline worker/awsCertlint/awsCertlint.go]: s/categoriezes/categorize/, but even though that sentence doesn't make a lot of sense. can you rephrase?

File: worker/awsCertlint/awsCertlint.go (hunk 1 above)
Comment (jvehent) [reply worker/awsCertlint/awsCertlint.go]: This should be configurable, similar to what we did with evChecker https://github.com/mozilla/tls-observatory/blob/master/worker/evCheckerWorker/evCheckerWorker.go#L24

File: worker/awsCertlint/awsCertlint.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	certlintDirectory = "/go/certlint" // path from tools/Dockerfile-scanner
+)
+
+type Result struct {
+	Bugs []string `json:"Bugs"`
+	Informational []string `json:"Informational"`
+	Notices []string `json:"Notices"`
+	Warnings []string `json:"Warnings"`
+	Errors []string `json:"Errors"`
+	FatalErrors []string `json:"FatalErrors"`
```
Comment (jvehent) [inline worker/awsCertlint/awsCertlint.go]: nit:gofmt
nit: json keys should be lowercase

File: worker/awsCertlint/awsCertlint.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		result.Success = true
+		result.Result = bs
+	}
+	resChan <- result
+}
+
+func (e eval) runCertlint(cert certificate.Certificate) (*Result, error) {
+	tmp, err := ioutil.TempFile("", "awslabs-certlint")
+	if err != nil {
+		return nil, fmt.Errorf("error creating temp dir", cert.ID)
```
Comment (jvehent) [inline worker/awsCertlint/awsCertlint.go]: s/temp dir/temp file/

Comment (jvehent) [reply worker/awsCertlint/awsCertlint.go]: also you're missing a `%s` in the format string

File: worker/awsCertlint/awsCertlint.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	defer os.Remove(tmp.Name())
+	x509Cert, err := cert.ToX509()
+	if err != nil {
+		return nil, fmt.Errorf("error converting to x509.Certificate", cert.ID)
```
Comment (jvehent) [inline worker/awsCertlint/awsCertlint.go]: missing format string

File: worker/awsCertlint/awsCertlint.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	if err := ioutil.WriteFile(tmp.Name(), x509Cert.Raw, 0644); err != nil {
+		return nil, fmt.Errorf("error writing x509.Certificate to temp file, err=%v", err)
+	}
+
+	// Run certlint over certificate
+	cmd := exec.Command("ruby", "-I", "lib:ext", "bin/certlint", tmp.Name())
```
Comment (jvehent) [inline worker/awsCertlint/awsCertlint.go]: I like to run those with a timeout to prevent them from hanging. Take a look at https://github.com/mozilla/tls-observatory/blob/master/connection/retriever.go#L47-L56

File: worker/awsCertlint/awsCertlint.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	// attach stdout/stderr
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("error running awslabs/certlint, err=%v, out=%q", cert.ID, strings.TrimSpace(stderr.String()))
+	}
+
+	return e.parseResponse(stdout)
+}
+
+
+// From: https://github.com/awslabs/certlint#output
```
Comment (jvehent) [inline worker/awsCertlint/awsCertlint.go]: I wonder if certlint could return JSON output to make it easier to process it on our side. What do you think of that, @pzb?

File: worker/awsCertlint/awsCertlint.go [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+// AnalysisPrinter outputs the results from awslabs/certlint over a given certificate
+func (e eval) AnalysisPrinter(input []byte, printAll interface{}) (results []string, err error) {
+	var result Result
+	if err := json.Unmarshal(input, &result); err != nil {
+		return nil, fmt.Errorf("awsCertlint Worker: failed to parse results: err=%v", err)
+	}
+
+	// Build results for webview
+	results = append(results, "* awslabs/certlint Results")
```
Comment (jvehent) [inline worker/awsCertlint/awsCertlint.go]: nit: remove "Results"
It would be nice to have a summary on the first line, such as `* awslabs/certlint: no issues found` or `1 warning found`

**Comment by jvehent** [comment]
This is looking pretty good!

File: worker/awsCertlint/awsCertlint.go (hunk 8 above)
Comment (adamdecaf) [reply worker/awsCertlint/awsCertlint.go]: Fyi, I filed an issue about this too. 

https://github.com/awslabs/certlint/issues/63

**Comment by adamdecaf** [comment]
@jvehent this should be good to go.

**Comment by adamdecaf** [comment]
@jvehent good call. I forgot about doing that. It's fixed now.
